File: src/js/stendhal/entity/GoldenCauldron.ts
/***************************************************************************
 *                   (C) Copyright 2025 - PolanieOnLine                    *
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

import { PopupInventory } from "./PopupInventory";
import { MenuItem } from "../action/MenuItem";
import { GoldenCauldronComponent } from "../ui/component/GoldenCauldronComponent";
import { FloatingWindow } from "../ui/toolkit/FloatingWindow";
import { Paths } from "../data/Paths";
import { singletons } from "../SingletonRepo";

declare var marauroa: any;
declare var stendhal: any;

const WINDOW_ID = "golden-cauldron";
const SLOT_CONTENT = "content";
const MAX_DIST_TO_VIEW = 4;
const BREW_TIME_MS = 5 * 60 * 1000;

export class GoldenCauldron extends PopupInventory {
	override zIndex = 5000;
	override maxDistToView = MAX_DIST_TO_VIEW;

	private open = false;
	private state = 0;
	private status = "";
	private brewer?: string;
	private readyAt = 0;
	private readyInSeconds = 0;
	private readyStartedAt = 0;
	private window?: FloatingWindow;
	private component?: GoldenCauldronComponent;
	private requestOpen = false;

	constructor() {
		super();
		this.sprite = {
			filename: Paths.maps + "/tileset/item/pot/cauldron.png",
			height: 64,
			width: 64,
			offsetX: 0,
			offsetY: 0
		};
	}

	override set(key: string, value: any) {
		super.set(key, value);
		if (key === "open") {
			this.open = true;
			this.syncWindowState();
		} else if (key === "state") {
			this.state = parseInt(value, 10) || 0;
			this.updateSprite();
			this.updateMixAvailability();
		} else if (key === "status") {
			this.status = value || "";
			this.updateStatus();
		} else if (key === "brewer") {
			this.brewer = value;
			this.updateMixAvailability();
		} else if (key === "ready_at") {
			this.readyAt = typeof value === "number" ? value : parseInt(value, 10) || 0;
			this.normalizeReadyAt();
			this.updateStatus();
			this.updateMixAvailability();
		} else if (key === "ready_in") {
			this.readyInSeconds = parseInt(value, 10) || 0;
			this.normalizeReadyAt();
			this.updateStatus();
			this.updateMixAvailability();
		} else if (key === "ready_started_at") {
			this.readyStartedAt = typeof value === "number" ? value : parseInt(value, 10) || 0;
			this.normalizeReadyAt();
			this.updateStatus();
			this.updateMixAvailability();
		}
	}

	override unset(key: string) {
		super.unset(key);
		if (key === "open") {
			this.open = false;
			this.requestOpen = false;
			this.closeInventoryWindow();
		} else if (key === "state") {
			this.state = 0;
			this.updateSprite();
			this.updateMixAvailability();
		} else if (key === "status") {
			this.status = "";
			this.updateStatus();
		} else if (key === "brewer") {
			this.brewer = undefined;
			this.updateMixAvailability();
		} else if (key === "ready_at") {
			this.readyAt = 0;
			this.readyInSeconds = 0;
			this.readyStartedAt = 0;
			this.updateStatus();
			this.updateMixAvailability();
		} else if (key === "ready_in") {
			this.readyInSeconds = 0;
			this.normalizeReadyAt();
			this.updateStatus();
			this.updateMixAvailability();
		} else if (key === "ready_started_at") {
			this.readyStartedAt = 0;
			this.normalizeReadyAt();
			this.updateStatus();
			this.updateMixAvailability();
		}
	}

	override buildActions(list: MenuItem[]) {
		super.buildActions(list);
		if (this.open) {
			list.push({
				title: "Przeszukaj",
				action: () => {
					this.requestOpen = true;
					this.openInventoryWindow();
				}
			});
			list.push({
				title: "Zamknij",
				action: () => {
					this.requestOpen = false;
					this.onUsed();
				}
			});
		} else {
			list.push({
				title: "Otwórz",
				action: () => {
					this.requestOpen = true;
					this.onUsed();
				}
			});
		}
	}

	override onclick(_x: number, _y: number) {
		if (this.open) {
			this.requestOpen = true;
			this.openInventoryWindow();
		} else {
			this.requestOpen = true;
			this.onUsed();
		}
	}

	private onUsed() {
		const action: { [key: string]: any } = {
			type: "use",
			zone: marauroa.currentZoneName,
			target_path: this.getIdPath()
		};
		if (!this.isContained()) {
			action.target = "#" + this["id"];
		}
		marauroa.clientFramework.sendAction(action);
	}

	private sendMixAction() {
		const action = {
			type: "goldencauldron",
			command: "mix",
			target_path: this.getIdPath(),
			zone: marauroa.currentZoneName
		};
		marauroa.clientFramework.sendAction(action);
	}

	private openInventoryWindow() {
		if (!this.open) {
			return;
		}
		if (!this.requestOpen && !this.isControlledByUser()) {
			return;
		}
		if (!this.window || !this.window.isOpen()) {
			const component = new GoldenCauldronComponent(
				this,
				SLOT_CONTENT,
				() => this.sendMixAction(),
				() => this.onWindowClosed()
			);
			component.updateStatus(this.status, this.readyAt);
			component.setMixEnabled(this.isControlledByUser() && !this.isActive());

			const state = stendhal.config.getWindowState(WINDOW_ID);
			const windowInstance = new FloatingWindow("Kocioł Draconii", component, state.x, state.y);
			windowInstance.setId(WINDOW_ID);
			windowInstance.enableMinimizeButton(true);
			windowInstance.setFixedWidth(232);
			this.window = windowInstance;
			this.component = component;
		}
		this.window.setMinimized(false);
		this.requestOpen = false;
		this.updateStatus();
		this.updateMixAvailability();
	}

	override closeInventoryWindow() {
		if (this.window && this.window.isOpen()) {
			this.window.close();
		}
		this.window = undefined;
		this.component = undefined;
	}

	private onWindowClosed() {
		this.window = undefined;
		this.component = undefined;
		this.requestOpen = false;
		if (this.open && this.isControlledByUser()) {
			this.onUsed();
		}
	}

	private syncWindowState() {
		if (this.open) {
			if (this.canViewContents()) {
				this.openInventoryWindow();
			}
		} else {
			this.closeInventoryWindow();
		}
	}

	private updateStatus() {
		if (this.component) {
			this.component.updateStatus(this.status, this.readyAt);
		}
	}

	private updateMixAvailability() {
		if (this.component) {
			const canMix = this.isControlledByUser() && !this.isActive();
			this.component.setMixEnabled(canMix);
		}
	}

	private isControlledByUser(): boolean {
		const name = singletons.getSessionManager().getCharName();
		if (!name || !this.brewer) {
			return false;
		}
		return name.toLowerCase() === this.brewer.toLowerCase();
	}

	private isActive(): boolean {
		return this.state > 0;
	}

	private updateSprite() {
		const offsetY = this.state > 0 ? 64 : 0;
		if (this.sprite) {
			this.sprite.offsetY = offsetY;
		}
	}

	private normalizeReadyAt() {
		let computed = this.readyAt;
		if (this.readyInSeconds > 0) {
			const candidate = Date.now() + this.readyInSeconds * 1000;
			if (computed > 0) {
				computed = Math.min(computed, candidate);
			} else {
				computed = candidate;
			}
		} else if (this.readyStartedAt > 0) {
			const expected = this.readyStartedAt + BREW_TIME_MS;
			if (computed <= 0 || computed > expected) {
				computed = expected;
			}
		}
		this.readyAt = computed;
	}
}

File: src/games/stendhal/client/entity/GoldenCauldron.java
/***************************************************************************
 *                   (C) Copyright 2003-2024 - Stendhal                    *
 ***************************************************************************/
/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.client.entity;

import games.stendhal.client.StendhalClient;
import games.stendhal.client.entity.User;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import marauroa.common.game.RPObject;
import marauroa.common.game.RPSlot;

/**
 * Client representation of the golden cauldron entity.
 */
public class GoldenCauldron extends Entity {
	private static final long BREW_TIME_MILLIS = 5L * 60L * 1000L;
	public static final Property PROP_OPEN = new Property();
	public static final Property PROP_STATE = new Property();
	public static final Property PROP_STATUS = new Property();
	public static final Property PROP_BREWER = new Property();
	public static final Property PROP_READY_AT = new Property();

	private boolean open;
	private int state;
	private String status;
	private String brewer;
	private RPSlot content;
	private long readyAt;
	private int readyInSeconds;
	private long readyStartedAt;
	private String cacheKey;

	private static final Map<String, Long> READY_AT_CACHE = new ConcurrentHashMap<String, Long>();
	private static final Map<String, Long> READY_STARTED_CACHE = new ConcurrentHashMap<String, Long>();

	public GoldenCauldron() {
		status = "";
	}

	public RPSlot getContent() {
		return content;
	}

	public boolean isOpen() {
		return open;
	}

	public boolean isActive() {
		return state > 0;
	}

	public String getStatusText() {
		return status;
	}

	public long getReadyAt() {
		return readyAt;
	}

	public boolean isControlledByUser() {
		final User user = User.get();
		if (user == null) {
			return false;
		}
		return brewer != null
		&& brewer.equalsIgnoreCase(StendhalClient.get().getCharacter());
	}

	@Override
	public void initialize(final RPObject object) {
		super.initialize(object);

		cacheKey = buildCacheKey(object);

		if (object.hasSlot("content")) {
			content = object.getSlot("content");
		} else {
			content = null;
		}

		open = object.has("open");
		state = object.has("state") ? object.getInt("state") : 0;
		status = object.has("status") ? object.get("status") : "";
		brewer = object.has("brewer") ? object.get("brewer") : null;
		readyAt = object.has("ready_at") ? object.getLong("ready_at") : 0L;
		readyStartedAt = object.has("ready_started_at") ? object.getLong("ready_started_at") : 0L;
		readyInSeconds = object.has("ready_in") ? object.getInt("ready_in") : 0;
		restoreCachedTimes();
		ensureReadyAtFromStart();
		applyReadyInSeconds();
		cacheReadyTimes();
	}

	@Override
	public void onChangedAdded(final RPObject object, final RPObject changes) {
		super.onChangedAdded(object, changes);

		if (changes.has("open")) {
			open = true;
			fireChange(PROP_OPEN);
		}
		if (changes.has("state")) {
			state = changes.getInt("state");
			fireChange(PROP_STATE);
		}
		if (changes.has("status")) {
			status = changes.get("status");
			fireChange(PROP_STATUS);
		}
		if (changes.has("brewer")) {
			brewer = changes.get("brewer");
			fireChange(PROP_BREWER);
		}
		if (changes.has("ready_at")) {
			readyAt = changes.getLong("ready_at");
			cacheReadyTimes();
			if (readyInSeconds <= 0) {
				fireChange(PROP_READY_AT);
			}
		}
		if (changes.has("ready_started_at")) {
			readyStartedAt = changes.getLong("ready_started_at");
			cacheReadyTimes();
			ensureReadyAtFromStart();
			if (!changes.has("ready_at") && !changes.has("ready_in")) {
				fireChange(PROP_READY_AT);
			}
		}
		if (changes.has("ready_in")) {
			readyInSeconds = changes.getInt("ready_in");
			applyReadyInSeconds();
			cacheReadyTimes();
			fireChange(PROP_READY_AT);
		}
	}

	@Override
	public void onChangedRemoved(final RPObject object, final RPObject changes) {
		super.onChangedRemoved(object, changes);

		if (changes.has("open")) {
			open = false;
			fireChange(PROP_OPEN);
		}
		if (changes.has("state")) {
			state = 0;
			fireChange(PROP_STATE);
		}
		if (changes.has("status")) {
			status = "";
			fireChange(PROP_STATUS);
		}
		if (changes.has("brewer")) {
			brewer = null;
			fireChange(PROP_BREWER);
		}
		if (changes.has("ready_at")) {
			readyAt = 0L;
			if (!changes.has("ready_in")) {
				readyInSeconds = 0;
			}
			cacheReadyTimes();
			fireChange(PROP_READY_AT);
		}
		if (changes.has("ready_in")) {
			readyInSeconds = 0;
			if (!changes.has("ready_at")) {
				fireChange(PROP_READY_AT);
			}
		}
		if (changes.has("ready_started_at")) {
			readyStartedAt = 0L;
			cacheReadyTimes();
			if (!changes.has("ready_at") && !changes.has("ready_in")) {
				fireChange(PROP_READY_AT);
			}
		}
	}

	private void restoreCachedTimes() {
		if (readyAt <= 0) {
			final Long cached = READY_AT_CACHE.get(getCacheKey());
			if ((cached != null) && (cached > System.currentTimeMillis())) {
				readyAt = cached.longValue();
			}
		}

		if (readyStartedAt <= 0) {
			final Long cachedStart = READY_STARTED_CACHE.get(getCacheKey());
			if (cachedStart != null) {
				readyStartedAt = cachedStart.longValue();
			} else if (readyAt > 0) {
				readyStartedAt = readyAt - BREW_TIME_MILLIS;
			}
		}
	}

	private void cacheReadyTimes() {
		final String key = getCacheKey();

		if (readyAt > 0) {
			READY_AT_CACHE.put(key, Long.valueOf(readyAt));
		} else {
			final Long cached = READY_AT_CACHE.get(key);
			if ((cached != null) && (cached.longValue() <= System.currentTimeMillis())) {
				READY_AT_CACHE.remove(key);
			}
		}

		if (readyStartedAt > 0) {
			READY_STARTED_CACHE.put(key, Long.valueOf(readyStartedAt));
		} else {
			final Long cachedStart = READY_STARTED_CACHE.get(key);
			if (cachedStart != null) {
				final long finish = cachedStart.longValue() + BREW_TIME_MILLIS;
				if (finish <= System.currentTimeMillis()) {
					READY_STARTED_CACHE.remove(key);
				}
			}
		}
	}

	private String getCacheKey() {
		if (cacheKey == null) {
			final RPObject.ID id = getID();
			final String zone = (id != null) ? id.getZoneID() : "";
			final int px = (int) Math.round(getX());
			final int py = (int) Math.round(getY());
			cacheKey = zone + ':' + px + ':' + py;
		}
		return cacheKey;
	}

	private String buildCacheKey(final RPObject object) {
		final RPObject.ID id = object.getID();
		final String zone = (id != null) ? id.getZoneID() : "";
		final int px = object.has("x") ? object.getInt("x") : (int) Math.round(getX());
		final int py = object.has("y") ? object.getInt("y") : (int) Math.round(getY());
		final String name = object.has("name") ? object.get("name") : "";
		return zone + ':' + name + ':' + px + ':' + py;
	}

	private void applyReadyInSeconds() {
		if (readyInSeconds <= 0) {
			ensureReadyAtFromStart();
			return;
		}

		final long candidate = System.currentTimeMillis() + (readyInSeconds * 1000L);
		if (readyAt > 0) {
			readyAt = Math.min(readyAt, candidate);
		} else if (readyStartedAt > 0) {
			final long expected = readyStartedAt + BREW_TIME_MILLIS;
			if (expected > System.currentTimeMillis()) {
				readyAt = Math.min(expected, candidate);
			} else {
				readyAt = candidate;
			}
		} else {
			readyAt = candidate;
		}
	}

	private void ensureReadyAtFromStart() {
		if (readyStartedAt <= 0) {
			return;
		}

		final long expected = readyStartedAt + BREW_TIME_MILLIS;
		if (expected <= 0) {
			return;
		}

		if (readyAt <= 0 || readyAt > expected) {
			readyAt = expected;
		}
	}
}
